patch/llpcPatchResourceCollect.cpp

```diff
@@ -3501,6 +3506,250 @@ void PatchResourceCollect::ReassembleOutputExportCalls()
     }
 }
 
+// =====================================================================================================================
+// Scalarize final vertex stage outputs and FS inputs ready for packing.
+void PatchResourceCollect::ScalarizeForPacking(
+    Module* pModule)    // [in/out] Module
+{
+    // First gather the input/output calls that need scalarizing.
+    SmallVector<CallInst*, 4> vsOutputCalls;
+    SmallVector<CallInst*, 4> fsInputCalls;
+    for (Function& func : *pModule)
+    {
+        if (func.getName().startswith(LlpcName::InputImportGeneric) ||
+            func.getName().startswith(LlpcName::InputImportInterpolant))
+        {
+            // This is a generic (possibly interpolated) input. Find its uses in FS.
+            for (User* pUser : func.users())
+            {
+                auto pCall = cast<CallInst>(pUser);
+                if (m_pPipelineShaders->GetShaderStage(pCall->getFunction()) != ShaderStageFragment)
+                {
+                    continue;
+                }
+                // We have a use in FS. See if it needs scalarizing.
+                if (isa<VectorType>(pCall->getType()) || (pCall->getType()->getPrimitiveSizeInBits() == 64))
+                {
+                    fsInputCalls.push_back(pCall);
+                }
+            }
+        }
+        else if (func.getName().startswith(LlpcName::OutputExportGeneric))
+        {
+            // This is a generic output. Find its uses in the final vertex stage.
+            for (User* pUser : func.users())
+            {
+                auto pCall = cast<CallInst>(pUser);
+                if (m_pPipelineShaders->GetShaderStage(pCall->getFunction()) !=
+                      m_pContext->GetPipelineContext()->GetFinalVertexStage())
+                {
+                    continue;
+                }
+                // We have a use the final vertex stage. See if it needs scalarizing. The output value is
+                // always the final argument.
+                Type* pValueTy = pCall->getArgOperand(pCall->getNumArgOperands() - 1)->getType();
+                if (isa<VectorType>(pValueTy) || (pValueTy->getPrimitiveSizeInBits() == 64))
+                {
+                    vsOutputCalls.push_back(pCall);
+                }
+            }
+        }
+    }
+
+    // Scalarize the gathered inputs and outputs.
+    for (CallInst* pCall : fsInputCalls)
+    {
+        ScalarizeInputGeneric(pCall);
+    }
+    for (CallInst* pCall : vsOutputCalls)
+    {
+        ScalarizeOutputGeneric(pCall);
+    }
+}
+
+// =====================================================================================================================
+// Scalarize a generic input.
+// This is known to be an FS generic or interpolant input that is either a vector or 64 bit.
+void PatchResourceCollect::ScalarizeInputGeneric(
+    CallInst* pCall)  // [in] The llpc.input.import.generic or llpc.input.import.interpolant call
```

Comment:
Maybe revise the comment not to mention the call name
// [in] Call importing generic inputs that is to be scalarized

I think always avoid referencing detailed code identifier names in comments (unless we documenting an interface) because they are subject to changes.

You can make asserts if you want to make assumption
LLPC_ASSERT(pCall->getCalledFunction()->getName() == ...)